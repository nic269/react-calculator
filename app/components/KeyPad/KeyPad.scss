@import 'app/styles/theme';

.container {
  border-top: $border-width solid $border-secondary-color;
}

.row {
  border-bottom: $border-width solid $border-primary-color;
   &:last-child {
     border-bottom: 0;
   }
}

.key {
  display: inline-block;
  width: $key-width;
  height: $key-height;
  background-color: $normal-key-bg-color;
  color: $normal-key-text-color;
  border: 0;
  border-right: $border-width solid $border-primary-color;
  cursor: pointer;
  font-size: 30px;

  &:focus {
    box-shadow: none;
    outline: none;
  }

  &:hover {
    background-color: darken($color: $normal-key-bg-color, $amount: 3)
  }

  &:last-child {
    border-right: 0;
  }

  &.medium {
    width: $key-width*2;
  }

  &.large {
    width: $key-width*3;
  }

  &.operator,
  &.submit {
    background-color: $operator-key-bg-color;
    color: $operator-key-text-color;

    &:hover {
      background-color: darken($color: $operator-key-bg-color, $amount: 5)
    }
  }

  &.clear {
    color: $text-secondary-color;
  }
}
