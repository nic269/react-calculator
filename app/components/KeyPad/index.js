/**
 *
 * KeyPad
 *
 */

import React, { memo } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

import styles from './KeyPad.scss';

const generateRowId = key => `row_${key}`;

const KeyPad = ({ onChange, layout }) => {
  const onClickButton = key => () => onChange(key);
  return (
    <div className={styles.container}>
      {layout.map((row, idx) => (
        <div className={styles.row} key={generateRowId(idx)}>
          {row.map(item => (
            <button
              type="button"
              key={item.id}
              className={cx(styles.key, styles[item.size], styles[item.group])}
              onClick={onClickButton(item.key)}
              title={item.description}
            >
              {item.name}
            </button>
          ))}
        </div>
      ))}
    </div>
  );
};

KeyPad.propTypes = {
  onChange: PropTypes.func,
  layout: PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.shape({}))),
};

KeyPad.defaultProps = {
  onChange: undefined,
  layout: [[]],
};

export default memo(KeyPad);
