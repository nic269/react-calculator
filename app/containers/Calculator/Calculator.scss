@import 'app/styles/theme';

.container {
  background-color: $calculator-bg-color;
  border: $border-width solid $border-primary-color;
  color: $white;
  font-size: 40px;
  line-height: 1;
  text-align: right;
  width: calc((#{$key-width}*#{$num-of-keys-in-a-row}) + (#{$border-width}*2));
}
