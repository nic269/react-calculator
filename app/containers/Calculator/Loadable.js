/**
 *
 * Asynchronously loads the component for Caculator
 *
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
