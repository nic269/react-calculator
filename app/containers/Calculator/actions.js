/*
 *
 * Caculator actions
 *
 */

import {
  GET_RESULT,
  CLEAR_RESULT,
  DELETE_AG,
  UPDATE_FORMULA,
} from './constants';

export const getResult = () => ({
  type: GET_RESULT,
});

export const clearResult = () => ({
  type: CLEAR_RESULT,
});

export const deleteAg = () => ({
  type: DELETE_AG,
});

export const updateFormula = payload => ({
  type: UPDATE_FORMULA,
  payload,
});
