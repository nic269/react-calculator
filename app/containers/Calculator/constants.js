/*
 *
 * Calculator constants
 *
 */
import _flattenDeep from 'lodash/flattenDeep';

export const GET_RESULT = 'app/Calculator/GET_RESULT';
export const CLEAR_RESULT = 'app/Calculator/CLEAR_RESULT';
export const DELETE_AG = 'app/Calculator/DELETE_AG';
export const UPDATE_FORMULA = 'app/Calculator/UPDATE_FORMULA';

export const ADD_KEY = '+';
export const SUB_KEY = '-';
export const MUL_KEY = '*';
export const DIV_KEY = '/';
export const CLEAR_GROUP = 'clear';
export const DELETE_GROUP = 'delete';
export const OPERATOR_GROUP = 'operator';
export const SUBMIT_GROUP = 'submit';
export const NUMBER_GROUP = 'number';
export const OPERATION_KEYS = [ADD_KEY, SUB_KEY, MUL_KEY, DIV_KEY];
export const SUBMIT_KEYS = ['Enter', '='];
export const CLEAR_KEYS = ['Escape'];
export const DELETE_KEYS = ['Backspace'];
export const LAYOUT = [
  [
    {
      id: '1',
      name: 'AC',
      group: CLEAR_GROUP,
      size: 'medium',
      key: 'Escape',
      description: 'Clear all (or press Esc)',
    },
    {
      id: '1_1',
      name: '⇤',
      group: DELETE_GROUP,
      key: 'Backspace',
      description: 'Delete (or press Backspace)',
    },
    {
      id: '2',
      name: '/',
      group: OPERATOR_GROUP,
      key: DIV_KEY,
      description: 'Divide (or press /)',
    },
  ],
  [
    {
      id: '3',
      name: '1',
      group: NUMBER_GROUP,
      key: '1',
    },
    {
      id: '4',
      name: '2',
      group: NUMBER_GROUP,
      key: '2',
    },
    {
      id: '5',
      name: '3',
      group: NUMBER_GROUP,
      key: '3',
    },
    {
      id: '6',
      name: '*',
      group: OPERATOR_GROUP,
      key: MUL_KEY,
      description: 'Multiply (or press *)',
    },
  ],
  [
    {
      id: '7',
      name: '4',
      group: NUMBER_GROUP,
      key: '4',
    },
    {
      id: '8',
      name: '5',
      group: NUMBER_GROUP,
      key: '5',
    },
    {
      id: '9',
      name: '6',
      group: NUMBER_GROUP,
      key: '6',
    },
    {
      id: '10',
      name: '-',
      group: OPERATOR_GROUP,
      key: SUB_KEY,
      description: 'Subtract (or press -)',
    },
  ],
  [
    {
      id: '11',
      name: '7',
      group: NUMBER_GROUP,
      key: '7',
    },
    {
      id: '12',
      name: '8',
      group: NUMBER_GROUP,
      key: '8',
    },
    {
      id: '13',
      name: '9',
      group: NUMBER_GROUP,
      key: '9',
    },
    {
      id: '14',
      name: '+',
      group: OPERATOR_GROUP,
      key: ADD_KEY,
      description: 'Add (or press +)',
    },
  ],
  [
    {
      id: '15',
      name: '0',
      group: NUMBER_GROUP,
      size: 'large',
      key: '0',
    },
    {
      id: '16',
      name: '＝',
      group: SUBMIT_GROUP,
      key: '=',
      description: 'Equal (or press Enter)',
    },
  ],
];
export const SUPPORTED_KEYS = [
  ..._flattenDeep(LAYOUT).map(item => item.key),
  'Enter',
  'Backspace',
];
