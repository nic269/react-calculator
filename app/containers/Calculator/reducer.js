/*
 *
 * Calculator reducer
 *
 */
import produce from 'immer';
import {
  GET_RESULT,
  CLEAR_RESULT,
  DELETE_AG,
  UPDATE_FORMULA,
  NUMBER_GROUP,
  OPERATOR_GROUP,
  ADD_KEY,
  SUB_KEY,
  MUL_KEY,
  DIV_KEY,
} from './constants';

const calculate = (front, operator, back) => {
  const frontVal = parseFloat(front);
  const backVal = parseFloat(back);
  let result;
  switch (operator) {
    case ADD_KEY:
      result = frontVal + backVal;
      break;
    case SUB_KEY:
      result = frontVal - backVal;
      break;
    case MUL_KEY:
      result = frontVal * backVal;
      break;
    case DIV_KEY:
      result = frontVal / backVal;
      break;
    default:
      result = front;
      break;
  }

  return result.toString();
};

export const initialState = {
  front: '0',
  back: '',
  operator: '',
};

/* eslint-disable default-case, no-param-reassign */
const calculatorReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case GET_RESULT: {
        const calculated = calculate(
          draft.front,
          draft.operator,
          draft.back ? draft.back : draft.front,
        );
        draft.front = calculated;
        draft.back = initialState.back;
        draft.operator = initialState.operator;
        break;
      }
      case CLEAR_RESULT:
        return initialState;
      case DELETE_AG:
        if (draft.front && draft.operator && draft.back) {
          draft.back =
            draft.back.length > 1
              ? draft.back.slice(0, draft.back.length - 1)
              : '';
        } else if (draft.front && draft.operator && !draft.back) {
          draft.operator = '';
        } else if (draft.front && !draft.operator) {
          draft.front =
            draft.front.length > 1
              ? draft.front.slice(0, draft.front.length - 1)
              : initialState.front;
        }
        break;
      case UPDATE_FORMULA:
        if (action.payload.group === NUMBER_GROUP) {
          if (!draft.operator) {
            draft.front = `${
              draft.front === initialState.front ? '' : draft.front
            }${action.payload.value}`;
          } else {
            draft.back = `${draft.back}${action.payload.value}`;
          }
        }
        if (action.payload.group === OPERATOR_GROUP) {
          if (!draft.operator || !draft.back) {
            draft.operator = action.payload.value;
          } else {
            const calculated = calculate(
              draft.front,
              draft.operator,
              draft.back,
            );
            draft.front = calculated;
            draft.back = initialState.back;
            draft.operator = action.payload.value;
          }
        }
        break;
    }
  });

export default calculatorReducer;
