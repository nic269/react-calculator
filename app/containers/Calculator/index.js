/**
 *
 * Calculator
 *
 */

import React, { memo, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import { useInjectReducer } from 'utils/injectReducer';
import KeyPad from 'components/KeyPad';
import Screen from 'components/Screen';
import { getResult, clearResult, deleteAg, updateFormula } from './actions';
import { makeSelectResult } from './selectors';
import reducer from './reducer';
import {
  LAYOUT,
  SUPPORTED_KEYS,
  OPERATION_KEYS,
  CLEAR_KEYS,
  SUBMIT_KEYS,
  DELETE_KEYS,
  CLEAR_GROUP,
  DELETE_GROUP,
  SUBMIT_GROUP,
  NUMBER_GROUP,
  OPERATOR_GROUP,
} from './constants';
import styles from './Calculator.scss';

const getGroupOfKey = key => {
  if (CLEAR_KEYS.includes(key)) return CLEAR_GROUP;
  if (DELETE_KEYS.includes(key)) return DELETE_GROUP;
  if (SUBMIT_KEYS.includes(key)) return SUBMIT_GROUP;
  if (OPERATION_KEYS.includes(key)) return OPERATOR_GROUP;
  return NUMBER_GROUP;
};

export const Calculator = props => {
  const keyDownHandler = e => {
    if (SUPPORTED_KEYS.includes(e.key)) {
      e.preventDefault();
      props.calculatorHandler(e.key);
    }
  };

  useInjectReducer({ key: 'calculator', reducer });

  useEffect(() => {
    document.addEventListener('keydown', keyDownHandler, false);

    return () => {
      document.removeEventListener('keydown', keyDownHandler, false);
    };
  }, []);

  return (
    <div className={styles.container}>
      <Screen value={props.result} />
      <KeyPad layout={LAYOUT} onChange={props.calculatorHandler} />
    </div>
  );
};

Calculator.propTypes = {
  calculatorHandler: PropTypes.func.isRequired,
  result: PropTypes.string,
};

Calculator.defaultProps = {
  result: '0',
};

const mapStateToProps = createStructuredSelector({
  result: makeSelectResult(),
});

export const mapDispatchToProps = dispatch => ({
  calculatorHandler: key => {
    const group = getGroupOfKey(key);
    switch (group) {
      case CLEAR_GROUP:
        dispatch(clearResult());
        break;
      case DELETE_GROUP:
        dispatch(deleteAg());
        break;
      case SUBMIT_GROUP:
        dispatch(getResult());
        break;
      default:
        dispatch(
          updateFormula({
            group,
            value: key,
          }),
        );
        break;
    }
  },
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(Calculator);
