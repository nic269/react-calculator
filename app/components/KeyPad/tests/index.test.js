/**
 *
 * Tests for KeyPad
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render, fireEvent } from 'react-testing-library';

import { LAYOUT } from 'containers/Calculator/constants';
import KeyPad from '../index';

describe('<KeyPad />', () => {
  let props;
  beforeEach(() => {
    props = {
      layout: LAYOUT,
      onChange: jest.fn(),
    };
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<KeyPad />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(<KeyPad {...props} />);
    expect(firstChild).toMatchSnapshot();
  });

  it('Should trigger onChange when a button has been clicked', () => {
    const { getByText } = render(<KeyPad {...props} />);
    fireEvent.click(getByText('clear'));
    expect(props.onChange).toHaveBeenCalledWith('Escape');
  });
});
