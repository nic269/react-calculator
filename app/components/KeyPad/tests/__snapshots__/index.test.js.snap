// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<KeyPad /> Should render and match the snapshot 1`] = `
<div>
  <div>
    <button
      class=""
      title="Clear all (or press Esc)"
      type="button"
    >
      AC
    </button>
    <button
      class=""
      title="Delete (or press Backspace)"
      type="button"
    >
      ⇤
    </button>
    <button
      class=""
      title="Divide (or press /)"
      type="button"
    >
      /
    </button>
  </div>
  <div>
    <button
      class=""
      type="button"
    >
      1
    </button>
    <button
      class=""
      type="button"
    >
      2
    </button>
    <button
      class=""
      type="button"
    >
      3
    </button>
    <button
      class=""
      title="Multiply (or press *)"
      type="button"
    >
      *
    </button>
  </div>
  <div>
    <button
      class=""
      type="button"
    >
      4
    </button>
    <button
      class=""
      type="button"
    >
      5
    </button>
    <button
      class=""
      type="button"
    >
      6
    </button>
    <button
      class=""
      title="Subtract (or press -)"
      type="button"
    >
      -
    </button>
  </div>
  <div>
    <button
      class=""
      type="button"
    >
      7
    </button>
    <button
      class=""
      type="button"
    >
      8
    </button>
    <button
      class=""
      type="button"
    >
      9
    </button>
    <button
      class=""
      title="Add (or press +)"
      type="button"
    >
      +
    </button>
  </div>
  <div>
    <button
      class=""
      type="button"
    >
      0
    </button>
    <button
      class=""
      title="Equal (or press Enter)"
      type="button"
    >
      ＝
    </button>
  </div>
</div>
`;
